File: Medium/2025-08-17-0837-New21Game/solution.js
/**
 * @param {number} N
 * @param {number} K
 * @param {number} maxPts
 * @return {number}
 */
var new21Game = function(N, K, maxPts) {
    if (K === 0 || N >= K - 1 + maxPts) return 1.0;
    let dp = new Array(N + 1).fill(0);
    dp[0] = 1.0;
    let probability = 0.0, windowSum = 1.0;
    for (let i = 1; i <= N; i++) {
        dp[i] = windowSum / maxPts;
        if (i < K) windowSum += dp[i];
        else probability += dp[i];
        if (i >= maxPts) windowSum -= dp[i - maxPts];
    }
    return probability;
};

File: Medium/2025-08-17-0837-New21Game/solution.py
class Solution:
    def new21Game(self, N: int, K: int, maxPts: int) -> float:
        if K == 0 or N >= K - 1 + maxPts:
            return 1.0
        dp = [0.0] * (N + 1)
        dp[0] = 1.0
        probability, window_sum = 0.0, 1.0
        for i in range(1, N + 1):
            dp[i] = window_sum / maxPts
            if i < K:
                window_sum += dp[i]
            else:
                probability += dp[i]
            if i >= maxPts:
                window_sum -= dp[i - maxPts]
        return probability
